### PR TITLE
add ttypdb: create pdbs on the bases of ttys in pods.

### PR DESCRIPTION
**Pull request by umezawatakeshi**
Signed-off-by: UMEZAWA Takeshi <takeshi-umezawa@cybozu.co.jp>

### CODE REVIEW

File: ttypdb/src/common/http.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		orig:   orig,
+		logger: logger,
+	}
+}
+
+func (h *proxyHTTPHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+
+	prw := &proxyHTTPResponseWriter{
+		orig: rw,
```
Comment (morimoto-cybozu) [inline ttypdb/src/common/http.go:31]:
```suggestion
		ResponseWriter: rw,
```

File: ttypdb/src/common/http.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+type proxyHTTPHandler struct {
+	orig   http.Handler
```
Comment (morimoto-cybozu) [inline ttypdb/src/common/http.go:11]:
```suggestion
	http.Handler
```

File: ttypdb/src/controller/poll.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+		if foundPdb {
+			err := pdbInterface.Delete(ctx, pod.Name, metav1.DeleteOptions{})
+			if err != nil {
+				logger.Error("failed to delete PDB", zap.Error(err))
+			} else {
+				logger.Info("deleted PDB")
+			}
+		} else {
+			logger.Debug("PDB does not exist")
+		}
```
Comment (morimoto-cybozu) [inline ttypdb/src/controller/poll.go:142]: You might use the "return early" pattern.

File: ttypdb/src/sidecar/count.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+			statBytes, err := os.ReadFile(filepath.Join("/proc", name, "stat"))
+			if err != nil {
+				return err
+			}
+
+			// The 6th (0-origin) field is controlling tty device number.
+			// If it is "0", the process is not controlled.
+			stat := string(statBytes)
+			idx := strings.LastIndexByte(stat, ')')
```
Comment (morimoto-cybozu) [inline ttypdb/src/sidecar/count.go:38]: It would be simpler if we could use `fmt.Sscanf()`.
Do we need to care about a command name with spaces?

Comment (umezawatakeshi) [reply ttypdb/src/sidecar/count.go:38]: Yes, wee need to do so.
procps (`ps` command) do the same thing: https://gitlab.com/procps-ng/procps/-/blob/v4.0.4/library/readproc.c?ref_type=tags#L578-581

**Comment by umezawatakeshi** [comment]
I also added a care for implicit status in proxyHTTPResponseWriter.

**Review by morimoto-cybozu — APPROVED** [review]
LGTM

**Pull request merged**